### PR TITLE
webgpu: NaN refactor

### PR DESCRIPTION
**Pull request by qjia7**
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7352)

### CODE REVIEW

**Comment by qjia7** [comment]
@hujiajie As offline talked with you, please take a look. Thanks.

File: tfjs-backend-webgpu/src/binary_op_util.ts [hunk 1]
```diff
@@ -110,22 +108,14 @@ const LOGICAL_OR = 'return f32(a >= 1.0 || b >= 1.0);';
...
-  }
+  let isNaN = b == 0.;
   var resultTemp = a % b;
-  if ((a < 0. && b < 0.) || (a >= 0. && b > 0.)) {
-    return resultTemp;
-  } else {
-    return (resultTemp + b) % b;
-  }
+  resultTemp = select((resultTemp + b) % b, resultTemp,
+      (a < 0. && b < 0.) || (a >= 0. && b > 0.));
```
Comment (hujiajie) [inline tfjs-backend-webgpu/src/binary_op_util.ts:114]: What's the line breaking rule? Follow clang-format?

File: tfjs-backend-webgpu/src/binary_op_util.ts [hunk 2]
```diff
@@ -201,26 +178,54 @@ const PRELU_VEC4 = `
...
+        doOpSnippet = useVec4 ? NOT_EQUAL_VEC4 : NOT_EQUAL;
+        break;
+      case BinaryOpType.POW:
+        doOpSnippet = useVec4 ? POW_VEC4 : POW;
+        break;
+      default:
+        continue;
+    }
+    const checkNaNSnippet =
+        useVec4 ? CHECK_NAN_SNIPPET_VEC4 : CHECK_NAN_SNIPPET;
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/binary_op_util.ts:213]: This is a bit strange to have a new variable whose name is the same as one of conditions. Can we just use "useVec4 ? CHECK_NAN_SNIPPET_VEC4 : CHECK_NAN_SNIPPET" directly?

**Review by gyagp — APPROVED** [review]
LGTM with one nit

File: tfjs-backend-webgpu/src/binary_op_util.ts [hunk 3]
```diff
@@ -201,26 +178,54 @@ const PRELU_VEC4 = `
...
-    return ${op}(a, b);
-  `;
-}
+const ATAN2 = 'var resultTemp = atan2(a, b);';
+const MAX = 'var resultTemp = max(a, b);';
+const MIN = 'var resultTemp = min(a, b);';
 
 export function getBinaryOpString(
     type: BinaryOpType, useVec4?: boolean): string {
+  // NaN-sensitive ops
```
Comment (hujiajie) [inline tfjs-backend-webgpu/src/binary_op_util.ts:187]: Perhaps "sensitive"/"insensitive" is not the right word here, but I'm not sure what could be better.

Comment (gyagp) [reply tfjs-backend-webgpu/src/binary_op_util.ts:187]: Maybe "Ops with NaN check" and "Ops without NaN check"?

**Review by qjia7 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7352)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @gyagp, @hujiajie, and @xhcao)

___
*[`tfjs-backend-webgpu/src/binary_op_util.ts` line 114 at r1](https://reviewable.io/reviews/tensorflow/tfjs/7352#-NNjsW7bEV5qUr-0PwcX-r1-114:-NNt8nry2d8CJsKdRHue:bu9enc8) ([raw file](https://github.com/tensorflow/tfjs/blob/c76722480e17f94b63634dd236ba4a225df236a1/tfjs-backend-webgpu/src/binary_op_util.ts#L114)):*
<details><summary><i>Previously, hujiajie (Jiajie Hu) wrote…</i></summary><blockquote>

What's the line breaking rule? Follow clang-format?
</blockquote></details>

I break it manually since it's too long. 
___
*[`tfjs-backend-webgpu/src/binary_op_util.ts` line 187 at r1](https://reviewable.io/reviews/tensorflow/tfjs/7352#-NNjsW7bEV5qUr-0PwcX-r1-187:-NNt9C344hDuwcpO6QmO:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/c76722480e17f94b63634dd236ba4a225df236a1/tfjs-backend-webgpu/src/binary_op_util.ts#L187)):*
<details><summary><i>Previously, gyagp (Yang Gu) wrote…</i></summary><blockquote>

Maybe "Ops with NaN check" and "Ops without NaN check"?
</blockquote></details>

Done.
___
*[`tfjs-backend-webgpu/src/binary_op_util.ts` line 213 at r1](https://reviewable.io/reviews/tensorflow/tfjs/7352#-NNjsW7bEV5qUr-0PwcX-r1-213:-NNt9DHz4jva9nv-IjuR:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/c76722480e17f94b63634dd236ba4a225df236a1/tfjs-backend-webgpu/src/binary_op_util.ts#L213)):*
<details><summary><i>Previously, gyagp (Yang Gu) wrote…</i></summary><blockquote>

This is a bit strange to have a new variable whose name is the same as one of conditions. Can we just use "useVec4 ? CHECK_NAN_SNIPPET_VEC4 : CHECK_NAN_SNIPPET" directly?
</blockquote></details>

Done.

**Review by gyagp — APPROVED** [review]

**Comment by hujiajie** [comment]
Thanks but sorry my brain bandwidth is a little exhausted, since the patch seems to be squeezing several changes into one. I would like to resort to others' feedback.

File: tfjs-backend-webgpu/src/binary_op_util.ts (hunk 2 above)
Comment (hujiajie) [reply tfjs-backend-webgpu/src/binary_op_util.ts]: While it's not the case here, sometimes clang-format behaves poorly on `${expression}`. To reiterate [my previous point](https://github.com/tensorflow/tfjs/pull/7193#discussion_r1053949182), personally I prefer the `${identifier}` style, or macro expansion in terms of C/C++.

Comment (gyagp) [reply tfjs-backend-webgpu/src/binary_op_util.ts]: clang-format behaves poorly on ${}, no matter it's expression or identifier. We can now leave it as is.

**Review by xhcao — APPROVED** [review]
LGTM

**Pull request merged**